Gems/Atom/RPI/Code/Source/RPI.Public/VisibleObjectContext.cpp

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright (c) Contributors to the Open 3D Engine Project.
+ * For complete copyright and license terms please see the LICENSE at the root of this distribution.
+ *
+ * SPDX-License-Identifier: Apache-2.0 OR MIT
+ *
+ */
+#include <Atom/RPI.Public/VisibleObjectContext.h>
+#include <Atom/RPI.Reflect/Base.h>
+#include <AzCore/Debug/Profiler.h>
+#include <AzCore/std/sort.h>
+
+AZ_DECLARE_BUDGET(RPI);
+
+namespace AZ
+{
+    namespace RPI
+    {
+        void VisibleObjectContext::Shutdown()
+        {
+            m_visibleObjectListContext.Clear();
+
+            m_finalizedVisibleObjectList.clear();
+        }
+
+        void VisibleObjectContext::AddVisibleObject(const void* userData, float depth)
+        {
+            if (userData)
+            {
+                VisibleObjectList& visibleObjectList = m_visibleObjectListContext.GetStorage();
+                VisibleObjectProperties visibleObject;
+                visibleObject.m_userData = userData;
+                visibleObject.m_depth = depth;
+                visibleObjectList.push_back(visibleObject);
+            }
+            else
+            {
+                AZ_Error(
+                    "VisibleObjectContext",
+                    false,
+                    "Null userData was added to a VisibleObjectContext. Visible object will be ignored.");
+            }
+        }
+
+        void VisibleObjectContext::FinalizeLists()
+        {
+            AZ_PROFILE_SCOPE(RPI, "VisibleObjectContext: FinalizeLists");
+            m_finalizedVisibleObjectList.clear();
```

Comment:
Suggestion: after clearing the finalized list, reserve enough space to hold the entire concatenation.